presto-main/src/main/java/com/facebook/presto/cost/UnnestStatsRule.java

```diff
@@ -0,0 +1,75 @@
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.cost.ComposableStatsCalculator.Rule;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.UnnestNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.plan.Patterns.unnest;
+import static java.lang.Double.NaN;
+
+public class UnnestStatsRule
+        implements Rule<UnnestNode>
+{
+    private static final int UPPER_BOUND_ROW_COUNT_FOR_ESTIMATION = 1;
+    @Override
+    public Pattern<UnnestNode> getPattern()
+    {
+        return unnest();
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> calculate(UnnestNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
+        // Since we don't have stats for cardinality about the unnest column, we cannot estimate the row count.
+        // However, when the source row count is low, the error would not matter much in query optimization.
+        // Thus we'd still populate the inaccurate numbers just so stats are populated to enable optimization
+        // potential.
+        PlanNodeStatsEstimate.Builder calculatedStats = PlanNodeStatsEstimate.builder();
+        if (sourceStats.getOutputRowCount() <= UPPER_BOUND_ROW_COUNT_FOR_ESTIMATION) {
+            calculatedStats.setOutputRowCount(sourceStats.getOutputRowCount());
+        }
+        else {
+            calculatedStats.setOutputRowCount(NaN);
+        }
+
+        for (Symbol symbol : node.getReplicateSymbols()) {
+            calculatedStats.addSymbolStatistics(symbol, sourceStats.getSymbolStatistics(symbol));
+        }
+        for (Map.Entry<Symbol, List<Symbol>> entry : node.getUnnestSymbols().entrySet()) {
+            List<Symbol> unnestToSymbols = entry.getValue();
+            SymbolStatsEstimate stats = sourceStats.getSymbolStatistics(entry.getKey());
+            for (Symbol symbol : unnestToSymbols) {
+                // This is a very conservative way on estimating stats after unnest. We assume each symbol
+                // after unnest would have as much data as the symbol before unnest. This would over
```

Comment:
If in understand it correctly, here we are estimating a column's average data size. And I'm estimating that each unnested column's average data size will be the same as the whole column before unnest. I think that's quite conservative regardless. It's a separate question whether the estimation is conservative or not after calculating #rows * average column size. But if that's off, it should be because the row count estimation is off.

Reply:
To clarify, I'm not sure we need to be conservative here.  why not keep the total data size the same and have the avg be `(total size before) / (#rows after)`.  I understand in general why be conservative for something that's uncertain, but given that we know the total data size can't grow, I don't think we gain anything from overestimating.

Reply:
Once we have stats for `#rows after` that would be the right thing to do. But since we don't have that yet, and there's no way of describing "I don't know how many rows there will be afterwards, but the total size should not change", being conservative is a good alternative. Even when we have stats for `#rows after`, we still need stats on each column (after unnest) to be able to have a more reasonable estimation. You can assume the average size of each row after unnest of an array is `total size / #rows`, but this does not work for map and row type, especially if you don't keep all columns after unnest. For example, for a `map<varchar, varchar>`, if I only keep one varchar column after unnest, we need to know the stats for keys and values separately to calculate the average size for both. Otherwise it's always safer to assume either one can be of the size of the whole.